### PR TITLE
[Feature] Add alert message and navigate screen when sign-in with google

### PR DESCRIPTION
**Pull request by devohno**
## Description

- When user sign in with google, we can get access token or user information on Expo or Native device env. After then, it'll shows alert message about that and move to [Main] screen.

- Documentation is already wrote, and it's good that I think. So I didn't create any modification.

## Related Issues

Related #27 

## Tests

I tested sign in flow with Expo and android native environment. But I didn't write test codes. I'm not fully understand about jest test, and interaction test with "sign in google" is not easy for me. Could I get some help with that? :)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk-mobile/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

**Comment by hyochan** [comment]
@devohno How is this going?

File: src/components/screen/Login.tsx [hunk 1]
```diff
@@ -156,6 +156,8 @@ function Screen(props: Props): React.ReactElement {
           }),
         });
         // console.log(response);
+        Alert.alert('login:' + JSON.stringify(response.accessToken));
+        onLogin();
```
[the remark this answers is not in the document]
Comment (devohno) [reply src/components/screen/Login.tsx]: Thanks for your opinion.
I used the `onLogin()` function that is defined in `Screen` component.
https://github.com/dooboolab/hackatalk-mobile/blob/8bd574fa04974dd43ad985837b6376c8af70cc09/src/components/screen/Login.tsx#L121

If I need to define a specific login function for google login, I'll refer to your opinion then. 👍

**Comment by devohno** [comment]
> @devohno How is this going?

It's soooooo late for this. I'm sorry.
There was some problem with testing. It seems to be a problem caused by lack of learning about testing. :(

I'll make this complete in this week!

**Comment by hyochan** [comment]
@Sandwichj Why do you approve this?

**Comment by devohno** [comment]
How can I cover the `codecov/patch` check?

**Comment by hyochan** [comment]
@devohno Try to run `yarn test --coverage` and see what's covered in last meetup from [youtube](https://www.youtube.com/watch?v=mVXCIcb4cbQ&t=165s).

**Comment by devohno** [comment]
@hyochan I checked where I have to covered from `codecov`

Screen Shot 2019-11-18 at 9 27 23 AM

I've tried to make mock for `Alert.alert`, but I couldn't yet.
When I pressed `btnGoogle`, it takes response from google API, and alert result message.
And navigate to `[Main]` screen.

Here's my test code by your guide.
```
// Login.test.tsx
import { Alert } from 'react-native';
...
it('should signin with [AppAuth] when ownership is expo', () => {
    jest.mock('Alert', () => {
      return {
        alert: jest.fn(),
      };
    });
    jest.mock('expo-constants', () => {
      return {
        appOwnership: 'expo',
      };
    });
    jest.mock('expo-app-auth', () => {
      return {
        authAsync: jest.fn(() => Promise.resolve()),
      };
    });

    const btnGoogle = testingLib.queryByTestId('btnGoogle');
    act(() => {
      fireEvent.press(btnGoogle);
    });
    expect(AppAuth.authAsync(null)).resolves.toBeCalled();
    expect(Alert.alert).toHaveBeenCalled();
    expect(props.navigation.navigate).toHaveBeenCalledWith('Main');
});
```
I got a fail from this section of test, showing "Cannot find module 'Alert' from 'Login.test.tsx'" is the error msg.

How can I cover `Alert` action?

**Comment by devohno** [comment]
Another test code here
```
// Login.test.tsx
import 'react-native';
...
it('should signin with [AppAuth] when ownership is expo', () => {
    ...
    const Alert = require('react-native');
    Alert.alert = jest.fn();

    const btnGoogle = testingLib.queryByTestId('btnGoogle');
    act(() => {
      fireEvent.press(btnGoogle);
    });
    expect(AppAuth.authAsync(null)).resolves.toBeCalled();
    expect(Alert.alert).toHaveBeenCalled();   // here's what I need to test!!
    expect(props.navigation.navigate).toHaveBeenCalledWith('Main');
});
```

But I got fail.
```
expect(jest.fn()).toHaveBeenCalled()

Expected number of calls: >= 1
Received number of calls:    0
```

**Comment by hyochan** [comment]
Little more detail to your code. A great try and a few lessons can be provided in your trial.

1. Try to mock `appOwnership` like below.
    ```
    import Constants, { AppOwnership } from 'expo-constants';

    jest.mock('expo-constants', () => ({
      ...jest.requireActual('expo-constants'),
      appOwnership: 'not-expo',
    }));
    ```

    Then, you can dynamically change the variable in `it` statement.
    `Constants.appOwnership = AppOwnership.Expo;`

    Since you are dynamically mocking the already imported module, below won't work.
    image


2. It'd be safer to mock implementation for `Alert` in `it` statement.
    ```
     jest.spyOn(Alert, 'alert').mockImplementation(() => jest.fn());
    ```

3. You're mocking solution is wrong. If you want to mock a `module` in `__mock__` dir you should use the same name like `expo-google-sign-in.ts` and then mock implementation.
    image


Since you seem to have considerable trouble making a `test` case for this, I'd like to push a commit that will help you understand what I've meant above.

**Comment by hyochan** [comment]
@devohno Since it was bit hard to point you out how to cover the test coverage, I've just done it myself. I hope you to follow the [commit](https://github.com/dooboolab/hackatalk-mobile/pull/43/commits/332ad3d180e9cb7b837c8a9bcce84e52ba0f00ec) and see how I've made it. Also, there still needs some improvement.
```
    // @ts-ignore
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve): void => {
        resolve({
          json: function() {
            return { id: 1 };
          },
        });
      });
    });
```

You may try to find how to mock above method without global.

File: src/components/screen/Login.tsx [hunk 2]
```diff
@@ -168,9 +169,12 @@ function Screen(props: Props): React.ReactElement {
...
       if (type === 'success') {
         setGoogleUser(user);
+        Alert.alert('login:' + JSON.stringify(user));
+        onLogin();
       }
-    } catch ({ message }) {
-      Alert.alert('login: Error:' + message);
+    } catch (error) {
+      // Alert.alert('login: Error:' + message);
+      Alert.alert(error);
```
Comment (hyochan) [inline src/components/screen/Login.tsx]: Try to change this to `Alert.alert(error.message)` as it is in `facebook`.

Comment (devohno) [reply src/components/screen/Login.tsx]: Alright. I'll try!

**Comment by hyochan** [comment]
Try to read the [article](https://jestjs.io/docs/en/asynchronous) in jest and try using `expect.assertions` in your case.

**Comment by devohno** [comment]
> @devohno Since it was bit hard to point you out how to cover the test coverage, I've just done it myself. I hope you to follow the [commit](https://github.com/dooboolab/hackatalk-mobile/pull/43/commits/332ad3d180e9cb7b837c8a9bcce84e52ba0f00ec) and see how I've made it. Also, there still needs some improvement.
> 
> ```
>     // @ts-ignore
>     global.fetch = jest.fn().mockImplementation(() => {
>       return new Promise((resolve): void => {
>         resolve({
>           json: function() {
>             return { id: 1 };
>           },
>         });
>       });
>     });
> ```
> 
> You may try to find how to mock above method without global.

I'm doing this code improvement now.

**Comment by devohno** [comment]
I tried to find how to mock `global.fetch` without `global`.
but there's no idea in my search.

But, I think if the mocking method about `global.fetch` apply globally, there will no repeatation for define to `global.fetch` in each of test.

And another way that I find is this.

`global.fetch = require('jest-fetch-mock');`

But the way is similar.

**Comment by hyochan** [comment]
> I tried to find how to mock `global.fetch` without `global`.
> but there's no idea in my search.
> 
> But, I think if the mocking method about `global.fetch` apply globally, there will no repeatation for define to `global.fetch` in each of test.
> 
> And another way that I find is this.
> 
> `global.fetch = require('jest-fetch-mock');`
> 
> But the way is similar.

I don't think this is a good idea. I've requested you to find solution not to approach globally but you've managed to set this into perfect global 😩 

However, I appreciate your work and I'll merge this since it has been quite a long time from the first `PR`.
I hope you to look for the [jest-fetch-mock](https://www.npmjs.com/package/jest-fetch-mock) and see how you can improve this code.

**Comment by hyochan** [comment]
@all-contributors please add @devohno for code, test

**Pull request merged**